Fix for number of used frame infos during crossgen2 compilation

gbalykov opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/59040

Problem with `JIT/Regression/JitBlue/GitHub_17777/GitHub_17777` happens because 1st `compCompile` invocation in `jitNativeCode` throws exception `CORJIT_INTERNALERROR`, but reserved unwind info (via `reserveUnwindInfo`) is never freed. Second attempt without opts succeeds, but `reserveUnwindInfo` is called again, so there's actually twice as much unwind info reserved as needed.

It might be worth to investigate the reason of `CORJIT_INTERNALERROR` on arm64 for GitHub_17777 separately. This change will still be needed, because compilation retry might happen on some other tests too for one reason or another.

Update: I've found the reason of `CORJIT_INTERNALERROR`. `NO_WAY` assert happens inside jit during compilation in emit.cpp:
```cpp
        /* Are we overflowing? */
        if (ig->igNext && (ig->igNum + 1 != ig->igNext->igNum))
        {
            NO_WAY("Too many instruction groups");
        }
```

`igNum` here is `unsigned` on all platforms, i.e. 32 bit. I guess this is currently a limitation of jit.

cc @alpencolt

gbalykov commented:
@MichalStrehovsky could you, please, take a look?

MichalStrehovsky commented:
@dotnet/crossgen-contrib could someone have a look. I'm not familiar with this.

jkotas commented on src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -422,6 +422,19 @@ private void PublishCode()
 
+            if (_usedFrameInfos != _numFrameInfos)
+            {
+                // jitNativeCode might retry itself when exception happens inside jit.
```
Would it be better to clear the already allocated frameinfos when jitNativeCode retries?

gbalykov replied:
Yes, you are right, it would be a better solution. At first I did not see existing jit interface to do this, but for some reason missed `reportFatalError`. I've checked that cleanup in `reportFatalError` works, so I'll update this PR.

gbalykov committed:
Fix for number of used frame infos during crossgen2 compilation

jitNativeCode might retry itself when exception happens inside jit,
but reserveUnwindInfo might be already called at this point,
so _numFrameInfos should be cleaned up.

gbalykov commented:
@jkotas could you, please, take a look?

jkotas commented on src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -3430,6 +3430,7 @@ private void reportFatalError(CorJitResult result)
         {
             // We could add some logging here, but for now it's unnecessary.
             // CompileMethod is going to fail with this CorJitResult anyway.
```
This comment is not correct. `CompileMethod` is not always going to fail.

jkotas commented on src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -3430,6 +3430,7 @@ private void reportFatalError(CorJitResult result)
         {
             // We could add some logging here, but for now it's unnecessary.
             // CompileMethod is going to fail with this CorJitResult anyway.
+            _numFrameInfos = 0;
```
Is this all state that needs to be cleaned up? From cursory look, there seems to be more state that may cause problems.

I am wondering whether the retry logic should be rather moved from the JIT to the EE side. It would allow the state cleanup to be more robust. Also, it would allow us to skip the retry in cases where it does not make sense. For example, it does not make sense to retry with no optimizations when generating Tier1 code.


@dotnet/jit-contrib Thoughts?

BruceForstall replied:
> I am wondering whether the retry logic should be rather moved from the JIT to the EE side.

Maybe I'm missing something: the proposed change is in crossgen2, not the JIT (or EE).

In the JIT'ing case, we already have a backout function: `CEEJitInfo::BackoutJitData()` => ` EEJitManager::RemoveJitData()`. It removes EH info, GC info, and maybe more. (Maybe not the debug info?) Seems like crossgen2 would need to do the same (i.e., remove GC info, debug info, code buffer).

jkotas replied:
We have similar method in crossgen2 as well (`CompileMethodCleanup`). The problem is that neither `BackoutJitData` nor `CompileMethodCleanup` gets called when the JIT decides to retry with optimizations off.

BruceForstall replied:
Ah, I see. There's a "small" window where, after the JIT has already started allocating EH/GC/code space from the VM, where it can hit a NO_WAY assert and decide to retry. It would make sense for the JIT to call back through the JIT/EE interface before retrying saying "I'm going to retry".

BruceForstall commented:
> `igNum` here is `unsigned` on all platforms, i.e. 32 bit. I guess this is currently a limitation of jit.

32 bits would be fine. The issue was with the prolog IG only, and should have been fixed by https://github.com/dotnet/runtime/pull/65153

mangod9 commented:
Hi @gbalykov, checking in whether you would be working on updating the PR?

trylek commented:
There has been no traffic on this PR for more than two months. I'm closing this now, please feel free to reopen as you see fit.
